src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1135,34 +1041,47 @@ public CachedConnection(HttpConnection connection)
         /// into the waiter as a result, and if no connection is available from the pool,
         /// this waiter's logic is used to create the connection.
         /// </summary>
-        private class ConnectionWaiter : TaskCompletionSource<(HttpConnectionBase connection, bool isNewConnection, HttpResponseMessage failureResponse)>
+        private sealed class ConnectionWaiter
         {
-            /// <summary>The pool with which this waiter is associated.</summary>
-            internal readonly HttpConnectionPool _pool;
-            /// <summary>Request to use to create the connection.</summary>
-            private readonly HttpRequestMessage _request;
-
-            /// <summary>Cancellation token for the waiter.</summary>
-            internal readonly CancellationToken _cancellationToken;
-            /// <summary>Registration that removes the waiter from the registration list.</summary>
-            internal CancellationTokenRegistration _cancellationTokenRegistration;
+            private readonly TaskCompletionSource<HttpConnection> _tcs;
+
             /// <summary>Next waiter in the list.</summary>
             internal ConnectionWaiter _next;
             /// <summary>Previous waiter in the list.</summary>
             internal ConnectionWaiter _prev;
 
             /// <summary>Initializes the waiter.</summary>
-            public ConnectionWaiter(HttpConnectionPool pool, HttpRequestMessage request, CancellationToken cancellationToken) : base(TaskCreationOptions.RunContinuationsAsynchronously)
+            public ConnectionWaiter()
             {
-                Debug.Assert(pool != null, "Expected non-null pool");
-                _pool = pool;
-                _request = request;
-                _cancellationToken = cancellationToken;
+                _tcs = new TaskCompletionSource<HttpConnection>(TaskCreationOptions.RunContinuationsAsynchronously);
             }
 
-            /// <summary>Creates a connection.</summary>
-            public ValueTask<(HttpConnection, HttpResponseMessage)> CreateConnectionAsync() =>
-                _pool.CreateHttp11ConnectionAsync(_request, _cancellationToken);
+            public async ValueTask<(HttpConnectionBase connection, bool isNewConnection, HttpResponseMessage failureResponse)> 
+                GetConnectionAsync(HttpConnectionPool pool, HttpRequestMessage request, CancellationToken cancellationToken)
+            {
+                CancellationTokenRegistration cancellationTokenRegistration = default;
+                if (cancellationToken.CanBeCanceled)
```

Comment:
This check probably doesn't hurt, but it also probably doesn't add much benefit; basically the first thing Register does is check CanBeCanceled.  So this could just be `CancellationTokenRegistration ctr = cancellationToken.Register(...);`, as long as, as my comment below suggests, the Register call doesn't result a delegate allocation.

Reply:
The code was previously doing this check, so I kept it. I'll rework.

Reply:
I'm not sure why I had that previously, but thanks for removing.  It's not only largely irrelevant, but the cancellation token is basically always going to be cancelable, because HttpClient always passes in a cancelable token, if for nothing else in support of CancelAllPending or whatever that method is called.  We'll onliy potentially get non-cancelable tokens when something other than HttpClient is used to access SocketsHttpHandler, so it's best to optimize for the cancelable rather than non-cancelable case.